### PR TITLE
Remove  Encode/Decode traits attempt 2

### PR DESCRIPTION
**Pull request by citizen-stig**
Proper version of https://github.com/Sovereign-Labs/sovereign/pull/257

### CODE REVIEW

File: sdk/src/state_machine/core/types.rs [hunk 1]
```diff
@@ -12,45 +10,18 @@ pub struct ArrayWitness {
...
-            next_idx: AtomicUsize::new(0),
-            hints: RefCell::new(hints),
-        })
-    }
-}
-
 impl Witness for ArrayWitness {
-    fn add_hint<T: crate::serial::Encode>(&self, hint: T) {
-        self.hints.borrow_mut().push(hint.encode_to_vec())
+    fn add_hint<T: BorshSerialize>(&self, hint: T) {
```
Comment (citizen-stig) [inline sdk/src/state_machine/core/types.rs:14]: This

File: sov-modules/sov-state/src/prover_storage.rs [hunk 2]
```diff
@@ -101,6 +101,7 @@ impl<S: MerkleProofSpec> Storage for ProverStorage<S> {
             let key_hash = KeyHash(S::Hasher::hash(key.key.as_ref()));
             // TODO: Switch to the batch read API once it becomes available
             let (result, proof) = untracked_jmt.get_with_proof(key_hash, latest_version)?;
+            // TODO: Question, wouldn't witness be "polluted" in case of error from below?
```
Comment (citizen-stig) [inline sov-modules/sov-state/src/prover_storage.rs:104]: This

Comment (preston-evans98) [reply sov-modules/sov-state/src/prover_storage.rs:104]: Hmm, yes I think it would. So maybe we should use `assert` instead of `ensure!`

Comment (citizen-stig) [reply sov-modules/sov-state/src/prover_storage.rs]: assert will panic. Fixed it with differnt approach

**Review by preston-evans98 — APPROVED** [review]
Looks great, thanks @citizen-stig!

**Review by bkolad — APPROVED** [review]

**Pull request merged**